core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type FileCache struct {
+	root    string
+	enabled bool
+}
+
+func NewFileCache(enabled bool) FileCache {
+	if !enabled {
+		return FileCache{enabled: false}
+	}
+	dir, found := os.LookupEnv("WANDB_CACHE_DIR")
+	if !found {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			slog.Error("Unable to find cache directory", "err", err)
+			return FileCache{enabled: false}
+		}
+		dir = filepath.Join(userCacheDir, "wandb")
+	}
+	return FileCache{root: filepath.Join(dir, "artifacts"), enabled: true}
+}
+
+// AddFile copies a file into the cache and returns the B64MD5 cache key.
+func (c *FileCache) AddFile(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return c.Write(f)
+}
+
+// AddFileAndCheckDigest copies a file into the cache. If a digest is provided, it also
+// verifies that the file's MD5 hash matches the digest.
+//
+// This is a convenience function intended for running in the background; instead of
+// returning an error it just logs it. This is often correct, since failing to add a
+// file to the cache shouldn't be a critical error.
+func (c *FileCache) AddFileAndCheckDigest(path string, digest string) {
+	if !c.enabled {
+		return
+	}
+	b64md5, err := c.AddFile(path)
+	if err != nil {
+		slog.Error("Unable to add file to cache", "err", err)
+	}
+	if digest != "" && digest != b64md5 {
+		slog.Error("File cache mismatch", "path", path, "expected", digest, "actual", b64md5)
+	}
+}
+
+// RestoreTo tries to restore the file referenced in a manifest entry to the given destination.
+//
+// If the file exists, it will be hashed and overwritten if the hash is different; if the
+// hash if correct, RestoreTo leaves it alone and returns true even if the cache is disabled.
+func (c *FileCache) RestoreTo(entry ManifestEntry, dst string) bool {
+	b64md5, err := utils.ComputeFileB64MD5(dst)
+	if err == nil && b64md5 == entry.Digest {
+		return true
+	}
+	if !c.enabled {
+		return false
+	}
+	cachePath, err := c.md5Path(entry.Digest)
+	if err != nil {
+		return false
+	}
+	exists, err := utils.FileExists(cachePath)
+	if !exists || err != nil {
+		return false
+	}
+	// TODO (hugh): should we set the LocalPath in the entry to the dst?
+	return utils.CopyFile(cachePath, dst) == nil
+}
+
+func (c *FileCache) md5Path(b64md5 string) (string, error) {
+	hexHash, err := utils.B64ToHex(b64md5)
+	if err != nil || !c.enabled {
+		return "", err
+	}
+	return filepath.Join(c.root, "obj", "md5", hexHash[:2], hexHash[2:]), nil
+}
+
+// Write copies the contents of the reader to the cache and returns the B64MD5 cache key.
+//
+// If the cache is disabled, the file will not be written to the cache but the hash will
+// still be computed and returned.
+func (c *FileCache) Write(src io.Reader) (string, error) {
+	if !c.enabled {
+		return copyWithHash(src, io.Discard)
+	}
+
+	tmpDir := filepath.Join(c.root, "tmp")
```

Comment:
should we do this during init? `Write` is a pretty hot path to do this over and over again?

Reply:
Yeah, it's reasonable to move the directory creation to the constructor; the reason I didn't is that constructors shouldn't panic or return nil, but I already have to deal with an error condition when we can't find a valid cache directory. I guess I also expect constructors to be lightweight.

We still have to create the destination directory, unless we want to create all 256 subdirectories at init time... which I could see, I suppose.